File: talks/ad-hoc-polymorphism/poly.py
class User(object):
    def __init__(self, contact_info):
        self.contact_info = contact_info

class Email(object):
    def __init__(self, email):
        self.email = email

    def contact(self):
        send_email(self.email)

class Phone(object):
    def __init__(self, phone):
        self.phone = phone

    def contact(self):
        send_text(self.text)

class Address(object):
    def __init__(self, address):
        self.address = address

    def contact(self):
        pass

File: talks/ad-hoc-polymorphism/Poly.hs
module Poly where

data ContactInfo
    = Email String
    | Phone String
    | Address String
    deriving (Show, Eq)

data User =
    User { contactInfo :: ContactInfo
         } deriving (Show, Eq)

type Total = Int

checkout :: Total -> User -> IO Bool
checkout total user = do
    successfulCharge <- charge total user
    if successfulCharge
       then contact (contactInfo user) >> return True
       else return False


contact :: ContactInfo -> IO ()
contact (Email s) = sendEmail s
contact (Phone phone) = sendText phone
contact (Address addr) = return ()

sendEmail = undefined
sendText = undefined
charge _ _ = return True
